src/theory/quantifiers/sygus/cegis_unif.cpp

```diff
@@ -0,0 +1,293 @@
+  {
+    Trace("cegis-unif-purify") << "PurifyFuncApp : candidate app " << nb
+                               << "\n";
+    for (i = 1; i < size; ++i)
+    {
+      if (!nb[i].isConst())
+      {
+        break;
+      }
+    }
+    if (i == size)
+    {
+      Node nv = d_parent->getModelValue(nb);
+      Trace("cegis-unif-purify") << "PurifyFuncApp : model value for " << nb
+                                 << " is " << nv << "\n";
+      /* test if different, then update model_guards */
+      if (nv != nb)
+      {
+        Trace("cegis-unif-purify") << "PurifyFuncApp : adding model eq\n";
+        model_guards.push_back(
+            NodeManager::currentNM()->mkNode(EQUAL, nv, nb).negate());
+        nb = nv;
+      }
+    }
+  }
+  /* put in cache */
+  Trace("cegis-unif-purify") << "... caching [" << n << "] = " << nb << "\n";
+  cache[n] = nb;
+  return nb;
+}
+
+Node CegisUnif::purifyLemma(
+    Node n,
+    std::vector<Node>& model_guards,
+    std::unordered_map<Node, Node, NodeHashFunction>& cache)
+{
+  bool fapp =
+      n.getKind() == APPLY_UF && n.getNumChildren() > 0 && n[0] == d_candidate;
+  /* top level applications of function-to-synthsize should never be removed */
+  if (!fapp)
+  {
+    std::unordered_map<Node, Node, NodeHashFunction>::const_iterator it =
+        cache.find(n);
+    if (it != cache.end())
+    {
+      Trace("cegis-unif-purify") << "... already visited " << n << "\n";
+      return it->second;
+    }
+  }
+  bool childChanged = false;
+  Kind k = n.getKind();
+  std::vector<Node> children;
+  /* Recurse */
+  for (unsigned i = 0, size = n.getNumChildren(); i < size; ++i)
+  {
+    Node child = fapp ? purifyFuncApp(n[i], model_guards, cache)
+                      : purifyLemma(n[i], model_guards, cache);
+    children.push_back(child);
+    childChanged = childChanged || child != n[i];
+  }
+  Node nb;
+  if (childChanged)
+  {
+    if (n.hasOperator())
+    {
+      children.insert(children.begin(), n.getOperator());
+    }
+    nb = NodeManager::currentNM()->mkNode(k, children);
+  }
+  else
+  {
+    nb = n;
+  }
+  /* top level applications of function-to-synthesize should not interfere with
+     internal applications are simplified */
+  if (!fapp)
+  {
+    /* put in cache */
+    Trace("cegis-unif-purify") << "... caching [" << n << "] = " << nb << "\n";
+    cache[n] = nb;
+  }
+  return nb;
+}
+
+void CegisUnif::registerRefinementLemma(const std::vector<Node>& vars,
+                                        Node lem,
+                                        std::vector<Node>& lems)
+{
+  Node plem;
+  std::vector<Node> model_guards;
+  std::unordered_map<Node, Node, NodeHashFunction> cache;
+  Trace("cegis-unif") << "Registering lemma at CegisUnif : " << lem << " with model value " << d_parent->getModelValue(lem) << "\n";
+  /* Make the purified lemma which will guide the unification utility. */
+  plem = purifyLemma(lem, model_guards, cache);
```

Comment:
I think we need to rewrite this Node.
Try purifying f( 1 + f( 0 ) ) = 0, this would be f( 0 ) != c V f( 1 + c ) = 0 for some constant c, but 1+c should be rewritten.
It would suffice to call plem = Rewriter::rewrite( plem ); here.